Make ProcessGroupAgent take num_threads as constructor argument

xush6528 opened the pull request:
Summary:
# Problem

If there is not enough number of thread in the RPC Agent thread pool. Some circular dependent works could cause deadlock.

The current to way to get around this deadlock is to provide abundant number of threads.

# Solution

as titled

Differential Revision: D17405491

mrshenli reviewed (commented):
Do you also need to add that to `init_model_parallel` as that's the entry point API?

zhaojuanmao commented on torch/distributed/rpc.py:
```diff
@@ -73,7 +74,7 @@ def _init_rpc(backend=RpcBackend.PROCESS_GROUP,
-        _agent = ProcessGroupAgent(self_name, group)
+        _agent = ProcessGroupAgent(self_name, group, num_send_recv_threads)
         init_rref_context(_agent)
     elif is_backend_registered(rpc_backend):
```
where is rpc_backend from? should it not be "backend"?

xush6528 replied:
@zhaojuanmao  Thanks. That's in @satgera 's PR. I am keeping each PR atomic.

xush6528 commented:
@mrshenli Right, updating it.

mrshenli commented on torch/distributed/__init__.py:
```diff
@@ -50,6 +51,6 @@ def init_model_parallel(self_name,
                 self_rank (int): a globally unique id/rank of this node.
                 init_method(str): backend specific init arguments.
             """
-            _init_rpc(backend, self_name, self_rank, init_method)
```
Add this to the doc `Arguments` section?

xush6528 committed:
Make ProcessGroupAgent take num_send_recv_threads as constructor argument (#26313)

Summary:
# Problem

If there is not enough number of thread in the RPC Agent thread pool. Some circular dependent works could cause deadlock.

The current to way to get around this deadlock is to provide abundant number of threads.

# Solution

as titled
Pull Request resolved: https://github.com/pytorch/pytorch/pull/26313

Differential Revision: D17405491

Pulled By: xush6528

fbshipit-source-id: e9b34db50b0ab614ebcb3414ae615ad1cde259ce
